apps/dataporten/views.py

```diff
@@ -0,0 +1,145 @@
+import logging
+
+from django.conf import settings
+from django.contrib import messages
+from django.contrib.auth.decorators import login_required
+from django.db import IntegrityError
+from django.shortcuts import redirect
+from oic import rndstr
+from oic.oauth2 import AuthorizationResponse, ResponseError
+
+from apps.dataporten.study.tasks import fetch_groups_information, find_user_study_and_update
+
+from .client import client_setup
+
+logger = logging.getLogger(__name__)
+
+DATAPORTEN_CLIENT_ID = settings.DATAPORTEN.get('STUDY', {}).get('CLIENT_ID')
+DATAPORTEN_CLIENT_SECRET = settings.DATAPORTEN.get('STUDY', {}).get('CLIENT_SECRET')
+DATAPORTEN_REDIRECT_URI = settings.DATAPORTEN.get('STUDY', {}).get('REDIRECT_URI')
+DATAPORTEN_SCOPES = settings.DATAPORTEN.get('STUDY', {}).get('SCOPES')
+
+
+@login_required()
+def study(request):
+    if request.user.is_member:
+        messages.info(request, 'Du er allerede registrert som medlem.')
+        return redirect('profiles_active', active_tab='membership')
+
+    logger.debug(
+        '{} wants to automatically confirm study programme through Dataporten.'.format(request.user),
+        extra={'user': request.user}
+    )
+
+    client = client_setup(DATAPORTEN_CLIENT_ID, DATAPORTEN_CLIENT_SECRET)
+
+    # Generate random values used to verify that it's the same user when in the callback.
+    state = rndstr()
+    nonce = rndstr()
+
+    request.session['dataporten_study_state'] = state
+    request.session['dataporten_study_nonce'] = nonce
+
+    args = {
+        'client_id': DATAPORTEN_CLIENT_ID,
+        'response_type': 'code',
+        'scope': DATAPORTEN_SCOPES,
+        'redirect_uri': DATAPORTEN_REDIRECT_URI,
+        'nonce': nonce,
+        'state': state,
+    }
+
+    logger.debug(
+        'Constructing authorization request and redirecting user to authorize through Dataporten.',
+        extra={'user': request.user}
+    )
+
+    auth_req = client.construct_AuthorizationRequest(request_args=args)
+    login_url = auth_req.request(client.authorization_endpoint)
+
+    return redirect(login_url)
+
+
+@login_required()
+def study_callback(request):
```

Comment:
Can this be split up somewhat?

Reply:
I tried, but I need access to the request for a lot of the messages, and both logging and messages end up spanning many lines. I'll look into it a bit more.